core/domain/exp_services.py

```diff
@@ -1710,6 +1710,516 @@ def is_voiceover_change_list(change_list):
+                elif change.content_id[:10] == 'rule_input':
+                    changed_property = 'answer_groups'
+                # A condition to store the name of the properties changed
+                # in changed_properties dict.
+                state_name = change.state_name
+                if state_name in new_to_old_state_names:
+                    state_name = new_to_old_state_names[change.state_name]
+                if state_name in changed_translations:
+                    if (changed_property not in
+                            changed_translations[state_name]):
+                        changed_translations[state_name].append(
+                            changed_property)
+                else:
+                    changed_translations[state_name] = [changed_property]
+                if state_name in changed_properties:
+                    if ('written_translations' not in
+                            changed_properties[state_name]):
+                        changed_properties[state_name].append(
+                            'written_translations')
+                else:
+                    changed_properties[state_name] = [
+                        'written_translations']
+
+        # Old_to_new_state_names: dict. Dictionary mapping state names of
+        # prev_exp_version to the state names of current_exp_version.
+        # It doesn't include the name changes of added/deleted states.
+        old_to_new_state_names = {
+            value: key for key, value in new_to_old_state_names.items()
+        }
+
+        if len(added_state_names) > 0 or len(deleted_state_names) > 0:
+            # Here we will send the changelist, version, latest_version,
+            # and exploration to the admin, so that the conditions
+            # can be reviewed.
+            return False
+
+        changes_are_mergeable = False
+
+        # state_names_of_renamed_states: dict. Stores the changes in
+        # states names in change_list where the key is the state name in
+        # frontend version and the value is the renamed name from the
+        # change list if there is any rename state change.
+        state_names_of_renamed_states = {}
+        for change in change_list:
+            change_is_mergeable = False
+            if change.cmd == exp_domain.CMD_RENAME_STATE:
+                old_state_name = change.old_state_name
+                new_state_name = change.new_state_name
+                if old_state_name in state_names_of_renamed_states:
+                    state_names_of_renamed_states[new_state_name] = (
+                        state_names_of_renamed_states.pop(old_state_name))
+                else:
+                    state_names_of_renamed_states[new_state_name] = (
+                        old_state_name)
+                if (state_names_of_renamed_states[new_state_name] not in
+                        old_to_new_state_names):
+                    change_is_mergeable = True
```

Comment:
Some of these codes looks duplicate. 

After taking a full pass, I think it would be better to have a separate ExplorationChangMergeVerifier class which will have the following properties:
```
ExplorationChangMergeVerifier:
  init(composite_change_list)
  state_name_change_map
  new_state_name
  changed_properties
  changed_translation etc.
  are_change_list_mergable(new_change_list_to_merge)
```
I'm not sure where we can reuse the ExplorationChangMergeVerifier for new_change_list_to_merge in the `are_change_list_mergable` function. I think that depends on the algorithm but I also see some duplicate fields and steps. I can help to finalize the class structure if needed.

I would suggest creating the new class in exp_domain or in a separate file.

At the end the exp_services.are_changes_mergeable should look something around:
```
Some basic check
return ExplorationChangMergeVerifier(compostie_change_list).are_change_list_mergeable(change_lis)
```

Reply:
I found we have an `ExplorationVersionsDiff` class you also use this class in the `are_change_list_mergeable` function if needed.

Reply:
Also, can we compare state-specific changes separately? i.e, there will be a base ExplorationChangeMergeVerifier which will internally use StateChangeMergeVerifier for cmd edit_state_property for all changes corresponding to a given state_name.

Reply:
Done.